Inlining

countvajhula opened the pull request:
### Summary of Changes

WIP inlining of separately-defined flows.

### Public Domain Dedication

- [x] In contributing, we relinquish any copyright claims on our contributions and freely release them into the public domain in the simple hope that they will provide value.

(**Why**: The freely released, copyright-free work in this repository represents an investment in a better way of doing things called _attribution-based economics_. Attribution-based economics is based on the simple idea that we gain more by giving more, not by holding on to things that, truly, we could only create because we, in our turn, received from others. As it turns out, an economic system based on attribution -- where those who give more are more empowered -- is significantly more efficient than capitalism while also being stable and fair (_unlike_ capitalism, on both counts), giving it transformative power to elevate the human condition and address the problems that face us today along with a host of others that have been intractable since the beginning. You can help make this a reality by releasing your work in the same way -- freely into the public domain in the simple hope of providing value. Learn more about attribution-based economics at [drym.org](https://drym.org), tell your friends, do your part.)

michaelballantyne commented:
With a little further thought though, I don't think you can both support the full mutual recursion behavior as exhibited in this program and get full inlining in all definition contexts, given the limitations of Racket's expansion. You could perhaps at module level with enough hacking, but I'm pretty sure it isn't possible in local definition contexts even with effects tricks. We can discuss more next time I join a Qi meeting.

eutro commented:
Yeah I was also thinking that there's no way our current solution works wrt mutually recursive definitions (thanks for the counterexample). I also had the thought, could it be better to store the surface syntax and then expand it only when we try to inline it? (Possibly caching expansions and, if we do them, compiler optimisations & analyses)? We can cut out the `#%host-expression` issues entirely and I don't see any semantics issues if the expansion of the form is otherwise pure enough. We settled on the approach we have now for reasons I don't recall (I think because we do want to store some processed version), but since that has fundamental issues, we can also just rederive the expansion when we need it from the surface syntax.

michaelballantyne commented:
One possible problem with storing unexpanded syntax is that you may expand the same syntax more than once, and thus execute side effects contained in macros more than once. Some bits of Racket do this sometimes, but generally it could create trouble. I'll have to think more deeply about whether there are other problems.

michaelballantyne commented:
Storing unexpanded syntax could also lead syntax to expand in a different module than where it is written, in the cross-module inlining case. One potential issue would be if that syntax includes host expressions with references to names with associated contracts, I think those references might expand in such a way that the associated blame would refer to the module into which they were inlined, which would be wrong.

countvajhula commented:
It sounds like we either pre-expand the syntax and undercut two-pass expansion, which totally breaks if there are mutually recursive flows, or we expand at inlining time and run afoul of an improper source module being inferred by the expander (which would affect contract blame, etc.). I wonder how the Racket compiler handles this kind of cross-module inlining. Or does it just not do that?

I suppose, in the second solution, of expanding first as part of inlining, it would be fine if we restrict it to just the current module, as we talked about last time. Would be nice if we could avoid that restriction eventually though.

Is there any way we could set some kind of parameter that would tell our invocation of `expand-flow` to consider that the expansion is occurring in a specified buffer (i.e., the source buffer, where the flow-to-be-expanded-and-inlined is defined)? Or could `expand-flow` accept a syntax object as "context" which could allow it to spoof the context of expansion?

michaelballantyne commented:
> I wonder how the Racket compiler handles this kind of cross-module inlining.

The Racket compiler does it as part of an additional pass after expansion. The problem is not that making it all work is in general impossible, but specifically that it is impossible _given the restriction to 2 passes_. And we're restricted to two passes because that's how many passes the Racket expander takes over definition contexts, and it doesn't give us the opportunity to take a 3rd (at least in local definition contexts).

Another possibility would be to restrict the form of mutual recursion that is allowed, while still allowing some. You could have a form with syntax:

```
(define-flows [name floe] ...)
```

where the flows defined in a single `define-flows` form can mutually-recur, but can otherwise only refer to things already declared above the `define-flows` form. This approach is easy to accomplish in `syntax-spec` via the [`host-interface/definitions`](https://docs.racket-lang.org/syntax-spec-v3/Specifying_languages.html#%28form._%28%28lib._syntax-spec-v3%2Fmain..rkt%29._host-interface%2Fdefinitions%29%29) form.

eutro committed:
scaffolding for inlining

This uses a transformer binding to capture the definition to be
inlined, and then does the inlining in a new compiler pass prior to
normalization.

(WIP from the meeting on Friday May 23 2025)

eutro committed:
get inlining to work by pre-expanding the captured syntax

(WIP from today's meeting, May 30 2025)

eutro committed:
try to fix 'disappeared-use -es

eutro committed:
Avoid recursion in inlining

In case of flows that contain references to themselves, for instance,
inlining would naively not terminate. This modifies the inlining pass
to signal to the `find-and-map` traversal *not* to traverse further
into the produced syntax.

(WIP from today's meeting)

eutro committed:
inline *all* non-recursive definitions

(WIP from today's meeting)

eutro committed:
delay define-flow expansion
